### PR TITLE
Issue-5: Add support for List extras

### LINKED ISSUE #5: Add support for List extras
Must include the following:

* The following static methods:
    * `Append :: [a] → [a] → [a]`: Appends two lists.
    * `FoldLeft :: Foldable t => (b → a → b) → b → t a → b`: Left-associative fold of a structure.
    * `FoldRight :: Foldable t => (a → b → b) → b → t a → b`: Right-associated fold of a structure.
    * `Head :: [a] → a`: Extracts the first element of a list which must be non-empty.
    * `Init :: [a] → [a]`: Extracts all of the elements before the last of a list which must be non-empty.
    * `IsEmpty :: Foldable a → Boolean`: Determines whether or not the list (read `Foldable`) is empty.
    * `Last :: [a] → a`: Extracts the last element of a list which must be non-empty.
    * `Length :: Foldable a → Number`: Returns the size/length of a finite structure.
    * `Map :: (a → b) → [a] → [b]`: Maps the underlying values within a list.
    * `Tail :: [a] → [a]`: Extracts the elements after the head of a list which must be non-empty.
    * `Uncons :: [a] → Maybe (a, [a])`: Decomposes a list into its head and tail. Returns `Nothing` if the list is empty; otherwise a `Just (x, xs)` where `x` is the head of the list and `xs` is the tail.

### PR DESCRIPTION
**Pull request by jlmorgan**
closes #5
SemVer:Minor
enhancement

### CODE REVIEW

**Review by doomchild — APPROVED** [review]
Seems legit.

File: docs/Data/Lists/FoldRight.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# `FoldRight<A, B>(fold, initialValue, list)`
+
+Folds a list from last to head.
```
Comment (JeremiahSanders) [inline docs/Data/Lists/FoldRight.md:3]: This workflow doesn't match the code. This matches `FoldLeft`.

File: docs/Data/Lists/Head.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+* `list (List<T>)`: The list.
+
+## Returns
+
+* `(T)`: First item in the list.
+
+## Throws
+
+* `ArgumentException` if the `list` is not a non-empty `List`.
```
Comment (JeremiahSanders) [inline docs/Data/Lists/Head.md]: I'm of two minds regarding the "is not a non-empty `List`" here. I like that it's consistent with line `3`, but the double-negative causes the mind to do a double-take when reading. Consider "is null or empty".

File: docs/Data/Lists/Init.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+* `ArgumentException` if the `list` is not a non-empty `List`.
+
+## Example
+
+```csharp
+Init(new List<int>());
+// => ArgumentException("list must be a not empty List")
+
+Init(new List<int> { 0 });
+// => List<int> { 0 }
```
Comment (JeremiahSanders) [inline docs/Data/Lists/Init.md:24]: Is this result correct? In a single-element list, isn't the first element also the last element, which line `3` says is excluded?

File: docs/Data/Lists/Init.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+Init(new List<int> { 1, 2, 3 });
+// => List<int> { 2, 3 }
```
Comment (JeremiahSanders) [inline docs/Data/Lists/Init.md]: Line `3` says the final item is omitted, but this looks to have the head omitted.

File: docs/Data/Lists/Length.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# `Length(list)`
+
+Gets the length of the `list`; otherwise, defaults to `0`.
```
Comment (JeremiahSanders) [inline docs/Data/Lists/Length.md:3]: Is the second half of the sentence needed (semicolon and on)?

File: docs/Data/Lists/Length.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+## Arguments
+
+* `list (List<T>)`: The list.
+
+## Returns
+
+* `(int)`: The length of the lifst.
```
Comment (JeremiahSanders) [inline docs/Data/Lists/Length.md]: "lifst" -> "list"

File: docs/Data/Lists/Head.md (hunk 2 above)
Comment (jlmorgan) [reply docs/Data/Lists/Head.md]: Hrm. I like you wording better.

File: docs/Data/Lists/FoldRight.md (hunk 1 above)
Comment (jlmorgan) [reply docs/Data/Lists/FoldRight.md:3]: Fixed.

File: docs/Data/Lists/Init.md (hunk 3 above)
Comment (jlmorgan) [reply docs/Data/Lists/Init.md:24]: That is correct.

Comment (jlmorgan) [reply docs/Data/Lists/Init.md:24]: Lists in functional scope are written as `[head, tail]`. If `tail` in this case does not exist, then `init` and `head` are effectively the same as lists are read head to tail and not tail to head.

File: docs/Data/Lists/Init.md (hunk 4 above)
Comment (jlmorgan) [reply docs/Data/Lists/Init.md]: Fixed.

File: docs/Data/Lists/Length.md (hunk 6 above)
Comment (jlmorgan) [reply docs/Data/Lists/Length.md]: Fixed.

File: docs/Data/Lists/Length.md (hunk 5 above)
Comment (jlmorgan) [reply docs/Data/Lists/Length.md:3]: The default is to acknowledge that the `list` may be `null`.

**Review by JeremiahSanders — APPROVED** [review]

**Pull request merged**